components/content_settings/core/browser/brave_content_settings_pref_provider.cc

```diff
@@ -635,39 +632,79 @@ void BravePrefProvider::UpdateCookieRules(ContentSettingsType content_type,
+    // oauth to work when the user sets custom overrides for a site.
+    // For example: Google OAuth will be allowed if the user allows all cookies
+    // and sets 3p cookie blocking for a site.
+    //
+    // We also create the same exception for firebase apps, since they
+    // are tightly bound to google, and require google auth to work.
+    // See: #5075, #9852, #10367
+    //
+    // PS: kGoogleLoginControlType preference might not be registered for tests.
     const auto google_auth_rule = Rule(
-        ContentSettingsPattern::FromString(kGoogleAuthPattern),
+        ContentSettingsPattern::FromString(permissions::kGoogleAuthPattern),
         ContentSettingsPattern::Wildcard(),
         ContentSettingToValue(CONTENT_SETTING_ALLOW),
         {.expiration = base::Time(), .session_model = SessionModel::Durable});
     rules.emplace_back(CloneRule(google_auth_rule));
     brave_cookie_rules_[incognito].emplace_back(CloneRule(google_auth_rule));
 
     const auto firebase_rule = Rule(
-        ContentSettingsPattern::FromString(kFirebasePattern),
+        ContentSettingsPattern::FromString(permissions::kFirebasePattern),
         ContentSettingsPattern::Wildcard(),
         ContentSettingToValue(CONTENT_SETTING_ALLOW),
         {.expiration = base::Time(), .session_model = SessionModel::Durable});
     rules.emplace_back(CloneRule(firebase_rule));
     brave_cookie_rules_[incognito].emplace_back(CloneRule(firebase_rule));
+  } else if (google_sign_in_pref_enabled) {
+    // Add per-site cookie exception for Google/Firebase auth domains.
+    // Get all sites that have BRAVE_GOOGLE_SIGN_IN turned on, and add exception
+    // for them
+    auto google_sign_in_content_setting_it = PrefProvider::GetRuleIterator(
+        ContentSettingsType::BRAVE_GOOGLE_SIGN_IN, incognito);
+    while (google_sign_in_content_setting_it &&
+           google_sign_in_content_setting_it->HasNext()) {
+      const auto google_sign_in_rule =
+          google_sign_in_content_setting_it->Next();
+      // The embedding pattern for the cookie rule will be the primary pattern
+      // for the BRAVE_GOOGLE_SIGN_IN permission
+      // We want to get all subdomains for the cookie rule
+      const auto embedding_pattern =
+          ContentSettingsPattern::ToDomainWildcardPattern(
+              google_sign_in_rule.primary_pattern);
+
+      const auto google_auth_rule = Rule(
+          ContentSettingsPattern::FromString(permissions::kGoogleAuthPattern),
+          embedding_pattern, google_sign_in_rule.value.Clone(),
+          {.expiration = google_sign_in_rule.metadata.expiration,
+           .session_model = google_sign_in_rule.metadata.session_model});
+      rules.emplace_back(CloneRule(google_auth_rule));
+      brave_cookie_rules_[incognito].emplace_back(CloneRule(google_auth_rule));
+
+      const auto firebase_rule = Rule(
+          ContentSettingsPattern::FromString(permissions::kFirebasePattern),
```

Comment:
@ShivanKaul please don't create these patterns on each rule iteration. Put them as `static const base::NoDestructor` near this block:
```
bool google_sign_in_flag_enabled = permissions::IsGoogleSignInFeatureEnabled();
```